File: Algorithm/src/chapter4/e25/e25.kt
package e25

import java.util.*

/**
 * Created by Yrh on 2015/12/12.
 */
fun main(args: Array<String>) {
    val scanner = Scanner(System.`in`)
    var a = Array(100, { IntArray(100) })
    var gain = IntArray(100)
    var q = FloatArray(100)
    var f = FloatArray(100)
    var temp = FloatArray(100)

    print("Input the number of item: ")
    val m = scanner.nextInt()

    print("Input how much money: ")
    val n = scanner.nextInt()

    print("Input one item gain table: ")
    for (j in 1..n) {
        q[j] = scanner.nextFloat()
        f[j] = q[j]
    }

    for (j in 1..n) {
        a[1][j] = j
    }

    for (k in 2..m) {
        print("Input another item gain table: ")
        for (j in 1..n) {
            temp[j] = q[j]
            q[j] = scanner.nextFloat()
            a[k][j] = 0
        }

        for (j in 0..n) {
            for (i in 0..j) {
                if (f[j - i] + q[i] > temp[j]) {
                    temp[j] = f[j - i] + q[i]
                    a[k][j] = i
                }
            }
        }

        for (j in 0..n) {
            f[j] = temp[j]
        }
    }

    var rest = n
    for (i in m downTo  1) {
        gain[i] = a[i][rest]
        rest -= gain[i]
    }

    for (i in 1..m) {
        print("${gain[i]} ")
    }

    print(f[n]);

}

File: Algorithm/src/chapter4/e25/e25.c
#include <stdio.h>
#include <memory.h>

int main()
{
	int i, j, k, m, n, rest, a[100][100], gain[100];
	float q[100], f[100], temp[100];
	
	memset(a, 0, sizeof(int) * 100 * 100);
	memset(gain, 0, sizeof(int) * 100);
	memset(q, 0, sizeof(float) * 100);
	memset(f, 0, sizeof(float) * 100);
	memset(temp, 0, sizeof(float) * 100);

	printf("How many item?");
	scanf("%d", &m);

	printf("How much money?");
	scanf("%d", &n);

	printf("Input one item gain table: ");
	for (j = 1; j <= n; j++)
	{
		scanf("%f", &q[j]);
		f[j] = q[j];
	}

	for (j = 0; j <= n; j++)
	{
		a[1][j] = j;
	}

	for (k = 2; k <= m; k++)
	{
		printf("Input another item gain table: ");
		for (j = 1; j <= n; j++)
		{
			temp[j] = q[j];
			scanf("%f", &q[j]);
			a[k][j] =  0;
		}
		for (j = 0; j <= n; j++)
		{
			for (i = 0; i <= j; i ++)
			{
				if (f[j - i] + q[i] > temp[j])
				{
					temp[j] = f[j - i] + q[i];
					a[k][j] = i;
				}
			}
		}
		for (j = 0; j <= n; j++)
		{
			f[j] = temp[j];
		}
	}

	rest = n;
	for (i = m; i >= 1; i --)
	{
		gain[i] = a[i][rest];
		rest -= gain[i];
	}
	for (i = 1; i <= m; i++)
	    printf("%d ", gain[i]);

	printf("\n%f", f[n]);
	return 0;
}
